File: Level 1/Cafeteria/java/Cafeteria.java
import java.util.*;

class Solution {
  
  public long getMaxAdditionalDinersCount(long N, long K, int M, long[] S) {
    Arrays.sort(S);
    
    long curStart = 1;
    long additionalDinersCount = 0;
    for (int i = 0; i < M; i++) {
      long curSeat = S[i];
      
      long freeSeats = (curSeat - K - curStart);
      if (freeSeats > 0) {
        additionalDinersCount += freeSeats / (K + 1);
        if (freeSeats % (K + 1) > 0) {
          additionalDinersCount++;
        }
      }
      
      curStart = curSeat + K + 1;
    }
    
    if (N + 1 > curStart) {
      additionalDinersCount += (N + 1 - curStart) / (K + 1);
      if ((N + 1 - curStart) % (K + 1) > 0) {
        additionalDinersCount++;
      }
    }
    
    return additionalDinersCount;
  }
  
}

File: Level 1/Cafeteria/python/Cafeteria.py
from typing import List
from math import ceil

def getMaxAdditionalDinersCount(N: int, K: int, M: int, S: List[int]) -> int:
    S.sort()

    additionalDiners = 0
    nextOccupiableSeat = 1
    for i in range(M):
        curSeat = S[i]

        additionalDiners += getMaxDinersBetweenSeats(nextOccupiableSeat, curSeat - K - 1, K)

        # Set nextOccupiableSeat to the next socially distanced seat after the current seat
        nextOccupiableSeat = curSeat + K + 1

    # Add the number of diners that can be seated between the last occupied seat and the last seat
    additionalDiners += getMaxDinersBetweenSeats(nextOccupiableSeat, N, K)

    return additionalDiners

# Determine the maximum number of diners that can be seated between two seat positions
def getMaxDinersBetweenSeats(startPos: int, endPos: int, K: int) -> int:
    num_seats = endPos - startPos + 1
    return ceil(num_seats / (K + 1)) if num_seats > 0 else 0